File: backend/app/schemas/referral.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional


class ReferralSettingsOut(BaseModel):
    reward_amount: float = Field(..., description="Flat reward amount in INR")
    is_active: bool
    reward_type: str = Field("PERCENTAGE", description="FLAT or PERCENTAGE")
    reward_percentage: float = Field(10.0, description="Percent of the first deposit paid to the referrer")
    min_deposit: float = Field(100.0, description="Minimum first deposit in INR that qualifies a referral")

    model_config = {"from_attributes": True}


class ReferralSettingsUpdateIn(BaseModel):
    reward_amount: float = Field(..., description="Flat reward amount in INR")
    is_active: bool
    reward_type: Optional[str] = Field(None, description="FLAT or PERCENTAGE")
    reward_percentage: Optional[float] = Field(None, ge=0, le=100, description="Percent of the first deposit")
    min_deposit: Optional[float] = Field(None, ge=0, description="Minimum qualifying first deposit in INR")


class ReferralStatsOut(BaseModel):
    referral_code: str
    referral_link: str
    reward_amount: float
    successful_referrals: int
    total_earnings: float
    pending_referrals: int
    reward_type: str = "PERCENTAGE"
    reward_percentage: float = 10.0
    min_deposit: float = 100.0


class ReferralHistoryOut(BaseModel):
    name: str
    username: str
    status: str
    reward_amount: float
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/services/referral.ts
import api from './api';

export interface ReferralStats {
  referral_code: string;
  referral_link: string;
  reward_amount: number;
  reward_type?: string;
  reward_percentage?: number;
  min_deposit?: number;
  is_active?: boolean;
  successful_referrals: number;
  total_earnings: number;
  pending_referrals: number;
}

export interface ReferralHistoryItem {
  name: string;
  username: string;
  status: 'COMPLETED' | 'PENDING';
  reward_amount: number;
  created_at: string;
}

export interface ReferralSettings {
  reward_amount: number;
  is_active: boolean;
  reward_type?: 'FLAT' | 'PERCENTAGE';
  reward_percentage?: number;
  min_deposit?: number;
}

export const referralService = {
  async getStats(): Promise<ReferralStats> {
    const res = await api.get('/referrals/stats');
    return res.data.data;
  },

  async getHistory(): Promise<ReferralHistoryItem[]> {
    const res = await api.get('/referrals/history');
    return res.data.data;
  },

  async getAdminSettings(): Promise<ReferralSettings> {
    const res = await api.get('/admin/referral/settings');
    return res.data.data;
  },

  async updateAdminSettings(payload: {
    reward_amount?: number;
    is_active?: boolean;
    reward_type?: 'FLAT' | 'PERCENTAGE';
    reward_percentage?: number;
    min_deposit?: number;
  } | number, is_active?: boolean): Promise<ReferralSettings> {
    const body = typeof payload === 'number'
      ? { reward_amount: payload, is_active: is_active ?? true }
      : payload;
    const res = await api.put('/admin/referral/settings', body);
    return res.data.data;
  },
};

/**
 * Human-readable summary of the active Refer & Win terms.
 * In PERCENTAGE mode the payout depends on the friend's first deposit, so a
 * flat rupee figure would misstate it.
 */
export function describeReferralReward(stats?: ReferralStats | null): {
  headline: string;
  perFriend: string;
  condition: string;
  isPercentage: boolean;
} {
  const isPercentage = (stats?.reward_type ?? 'PERCENTAGE') === 'PERCENTAGE';
  const pct = stats?.reward_percentage ?? 10;
  const minDeposit = stats?.min_deposit ?? 100;
  const flat = stats?.reward_amount ?? 100;

  return {
    isPercentage,
    headline: isPercentage ? `${pct}%` : `₹${flat}`,
    perFriend: isPercentage
      ? `${pct}% of their first deposit`
      : `₹${flat} per friend`,
    condition: `Minimum first deposit ₹${minDeposit}`,
  };
}
